cache/refs.go

```diff
@@ -906,14 +856,38 @@ func (sr *mutableRef) Mount(ctx context.Context, readonly bool, s session.Group)
 	return sr.mount(ctx, readonly)
 }
 
+// must be called holding cacheRecord mu
+func (sr *mutableRef) mount(ctx context.Context, readonly bool) (snapshot.Mountable, error) {
+	m, err := sr.cm.Snapshotter.Mounts(ctx, getSnapshotID(sr.md))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to mount %s", sr.ID())
+	}
+	if readonly {
+		m = setReadonly(m)
+	}
+	return m, nil
+}
+
 func (sr *mutableRef) Commit(ctx context.Context) (ImmutableRef, error) {
 	sr.cm.mu.Lock()
 	defer sr.cm.mu.Unlock()
 
 	sr.mu.Lock()
 	defer sr.mu.Unlock()
 
-	return sr.commit(ctx)
+	rec := sr.cacheRecord
+
+	ir := &immutableRef{
```

Comment:
Does this mean that the `ID()` does not change after commit or am I missing something. That would be problematic for the cache (and other queries) that work on IDs if they are not unique.

Reply:
Yes it does mean that, that's one of the main simplifications I think, now the underlying snapshot data is only ever represented by a single cache record rather than two different ones depending on whether it's being presented to the outside callers as mutable or immutable. So whereas previously `Get` and `GetMutable` would always return entirely different records for the same underlying snapshot data, now they can be called with the same ID to obtain either a immutable or mutable reference, respectively, to that record, which makes more sense IMO and simplifies the code quite a bit.

In terms of being backwards compatible, one important point is that the higher level solver cache never saved a `MutableRef`, it only ever saved `ImmutableRef`s. That meant that the migration logic I added in `cacheManager.init` just needed to handle finding immutable refs with `equalMutable` set and moving the snapshot and other metadata over to the now centralized record.

However, I did just realize there are also the various indexes scattered around and used to search for mutable records by git source, local source, exec's shared cache mounts, etc. Those call GetMutable with an ID found by searching the bolt db, so as of right now all those caches would get invalidated after upgrading to a buildkit with this change because those records would be something's `equalMutable` and thus removed as part of the migration logic in `cacheManager.init`.

I can fix that issue by just updating init to also migrate the indexes it finds on the `equalMutable` to the centralized record, which I will do.

Reply:
>  important point is that the higher level solver cache never saved a MutableRef, it only ever saved ImmutableRefs

Yes, but it shouldn't mean the `ImmutableRef` was guaranteed to be finalized. The simple case should be 1) get immutableref 2) save cache for it 3) it becomes mutable again and changes 4) after commit it is the same id for immutableref but different id 4) cache lookup returns the wrong record.

Reply:
I'm not sure I'm following where a wrong record gets returned by cache lookup. Are you assuming that the snapshot ID and ref ID are always the same? If so, that isn't the case with this change (and actually wasn't the case before this change, though in a smaller set of circumstances). The snapshot ID is now treated as just a piece of metadata tracked for the ref and can get updated during a call to finalize. In the case of a finalized ref, the snapshot ID and ref ID just so happen to actually be the same, but that's just because it made some crash recovery logic a bit simpler to write (see [here](https://github.com/sipsma/buildkit/blob/b2cbd4e0eb7dd189174659e740c8e6f36db07d85/cache/refs.go#L833-L834) and [here](https://github.com/sipsma/buildkit/blob/b2cbd4e0eb7dd189174659e740c8e6f36db07d85/cache/manager.go#L467-L468)), it's not strictly necessary or assumed anywhere else in the code.

So in the case I think you're describing, here's what happens:
1. You get an immutable ref, it has ID 123 and snapshotID 456. The snapshot is not committed.
2. You save that immutable ref for 123 in the cache somewhere and release the ref
3. You obtain a mutable ref by calling GetMutable with ID set to 123 and change the data on underlying mount for snapshot 456
4. You call `Commit` on the mutable ref for 123, turning it into an immutable ref and then finalize that immutable ref. During finalize, the record's snapshot ID is internally updated from 456 to 789
5. You later get an immutable ref for the record again by calling Get with ID set to 123. When you mount it, you are using snapshot 789 now because the record has migrated to using that finalized snapshot.

Overall, the tradeoff here is that now the snapshot ID for a cache record can change, but I think that makes much more conceptual sense than changing a ref's ID just because the snapshot data got committed to the driver. Callers outside the cache package shouldn't care at all about snapshot IDs, only ref IDs, so having them be aware and handle the fact ref IDs change just because some underlying snapshot ID that they don't know about change felt like it wasn't providing any benefit.

Reply:
Snapshot IDs are always internal, they don't matter to me. It is the ref ID that gets stored in the build cache boltdb https://github.com/moby/buildkit/tree/master/solver/bboltcachestorage . And yes there is also the case with the search for mutable that you mentioned where currently I think the metadata was different for these records, but I need to look at it more to understand that case.

Reply:
Okay, just pushed updates for your comments and a separate commit for migrating the indexes too. Let me know if it makes sense to you. AFAICT that should handle everything and make this change fully backwards compatible without  even invalidating any caches after upgrade. All of the current tests pass too of course, so I don't believe there's any cases where wrong records get returned.

I do have that simple test case for migrating the equalMutable record, but ideally there would be a way to test it truly end-to-end; something like start up an actual buildkitd v0.9 binary, run a bunch of solves to populate the cache and then start up the buildkitd being tested to verify all the cache gets migrated correctly without any invalidation. Am I correct that we can't do that today? Does something like that sound plausible to you to implement (and worth implementing)?

Reply:
> All of the current tests pass too of course, so I don't believe there's any cases where wrong records get returned.

I'm not sure if the cases are very practical (there are not a lot of places that call `GetMutable` but I think at least local source should be affected by it, maybe also exec mounts), but if there exists a case where there can be `ImmutableRef`  who has ID and then later there is another `ImmutableRef` that has the same ID but a different content then this does not satisfy the interface requirements.

> Am I correct that we can't do that today?

Yes, you can't run an older version of the daemon for tests atm. to test migrations. Not against this possibility but it is likely also quite messy to add.

Reply:
> if there exists a case where there can be ImmutableRef who has ID and then later there is another ImmutableRef that has the same ID but a different content then this does not satisfy the interface requirements.

I actually thought such cases already existed before this change and I also can't think of any reason why we would want to prevent them or make this be a requirement of the interface. To make sure we're on the same page, I'm not thinking of the case where two ImmutableRefs open at the same time with the same ref ID have different underlying snapshot data (that would be bad and shouldn't happen); just the case where an unfinalized immutable ref is open for a cache record, then later a mutable ref is open for the record and gets its snapshot mount changed and then after that an immutable ref gets opened again for the record (now with different underlying snapshot data from the first immutable ref).

That case only happens for immutable refs of cache records that have not been finalized, which means the record is not the parent of any other records and that it has not been exported in any way. For such records, what specific problems could be caused by the underlying data changing when it gets re-opened as mutable (provided the rules about simultaneous opening of immutable and mutable refs are honored)? I thought that was the whole point of the lazy-commit optimization; that you can delay committing to the underlying driver until actually necessary for the purposes of consistency (such as when a record is going to become the parent of another one or going to be exported).

Basically, all the situations in which changing the underlying data changing for a given ref ID causes problems are situations in which finalize will have been called, which prevents the data from changing, right? Or am I missing something else that causes problems?

Reply:
> I actually thought such cases already existed before this change and I also can't think of any reason why we would want to prevent them or make this be a requirement of the interface.

I don't think this problem existed in the previous implementation cause I actually remember hitting these cases when developing this. It might have even been a reason for the `equal*` design.

> For such records, what specific problems could be caused by the underlying data changing when it gets re-opened as mutable (provided the rules about simultaneous opening of immutable and mutable refs are honored)? I thought that was the whole point of the lazy-commit optimization; 

Yes, the point of lazy commit is that data can become mutable again. But once it does, the immutable should become invalid. The `ID()` is how other components(like the build cache system, or incremental context send) query references and store them for persistence. This is what they use to understand if two records are equal.

Reply:
If you look at https://github.com/moby/buildkit/commit/eac79f7c7e936453a58508761638e16ffb255085 then previous implementation was based on `Freeze` logic. The ID not changing seems to be one of the main differences.

Reply:
It makes sense that the ID is persisted and used to lookup records and check whether records are equal, but my point is that the only scenarios I can think of where problems could theoretically arise are either impossible or they are scenarios in which the records will have been finalized (they become the parent of another record or they got exported), which prevents the problems from happening in the first place.

Can you give a specific example of a scenario where you think this would cause problems so I can write a test for it and see what the behavior is before and after this change?

I've been trying to write a test myself but honestly can't even find a way to trigger it. I was trying something along the lines of:
```
// create a local dir, which will be imported and then copied to scratch, which will be exported to a different local dir
localdir, err := tmpdir(
  fstest.CreateFile("foo", []byte("foo"), 0600),
  fstest.CreateFile("bar", []byte("bar"), 0600),
)

destdir, err := ioutil.TempDir("", "buildkit")

local := llb.Local("source")

def := llb.Scratch().
  File(llb.Copy(local, "/", "/")).Marshal(sb.Context())

c.Solve(sb.Context(), def, SolveOpt{
  Exports: []ExportEntry{
	{
		Type:      ExporterLocal,
		OutputDir: destdir,
	},
  },
  LocalDirs: map[string]string{
	"source": localdir,
  },
}, nil)

// assert the exported dir has the expected contents
dt := ioutil.ReadFile(filepath.Join(destdir, "foo"))
require.Equal(t, []byte("foo"), dt)
dt = ioutil.ReadFile(filepath.Join(destdir, "bar"))
require.Equal(t, []byte("bar"), dt)

// change the local dir and then run the same solve again
os.WriteFile(filepath.Join(localdir, "bar"), []byte("doo"), 0600)

destdir = ioutil.TempDir("", "buildkit")

c.Solve(sb.Context(), def, SolveOpt{
Exports: []ExportEntry{
	{
		Type:      ExporterLocal,
		OutputDir: destdir,
	},
  },
  LocalDirs: map[string]string{
	"source": localdir,
  },
}, nil)

// assert the change happened as expected
dt = ioutil.ReadFile(filepath.Join(destdir, "foo"))
require.Equal(t, []byte("foo"), dt)
dt = ioutil.ReadFile(filepath.Join(destdir, "bar"))
require.Equal(t, []byte("doo"), dt)
```

But that can't actually reproduce the behavior we are talking about because [local cache keys are based on session IDs](https://github.com/moby/buildkit/blob/b2195cec759428a7775a6fe01799b961d14a81bc/source/local/local.go#L94), which means that the solver's cache won't ever try reloading a previous result for a local (so the test passed).

Then I tried the session re-use feature to force the cache keys to be the same for the local vertex, but the test just failed with errors like: `DeadlineExceeded: no active session for 7fa0r79idbckfsm9kjkcdgann: context deadline exceeded`. IIRC the shared session feature is known to be broken, so that's not a surprising error, right?

Either way, even if I was able to get the shared session to work, I still am pretty sure I wouldn't have been able to trigger the scenario we were talking about because it would have resulted in the local just getting reloaded rather than made mutable again (seems like a pre-existing bug hidden by the fact that shared sessions don't work, but not 100% sure).

So I'm at a loss for how it's possible to reproduce the problem you are mentioning, let me know if you have any ideas.

Reply:
As far as the LLB API is concerned caching via sessionid is a valid feature as it is the only way for the client to confirm that the files did not change without retransfer. Possibly `buildx bake` will start to rely on this and for more complex cases this is definitely useful.

Even if you make a case that there is little use for affected callers atm, implementing the interface in a way that methods can only be called in a certain order limits future development and increases chances for this to unexpectedly break. In that case, the interface should be changed so callers would use some other, more reliable method for equality and persistence use-cases.

Reply:
Okay, I can only imagine it breaking in situations where the rest of the code changes in pretty fundamental ways, but my imagination is limited here so that's alright.

Think I found a solution that retains the previous behavior of invalidating immutable ref IDs if they are reopened as mutable while still simplifying things relative to before. Namely, I changed it so that if you call `cacheManager.GetMutable` w/ the ID of a cache record that has been committed but not finalized, then you get a record that has the same metadata as before but with a new ID and the old ID is invalidated. This is implemented in `GetMutable` by just creating a new record and copying the metadata from the previous one, so it's pretty simple conceptually.

This seems to have been mostly seamless, with the one slight exception being the contenthash code, where I had to change from using the ref ID as key to using a metadata field in the cache record. This was very straightforward but let me know if there's any issues I'm not seeing as I'm not super familiar with that code.

Honestly, in the longer run, I suspect the best solution would be to replace the lazy-commit approach with something functionally equivalent that doesn't require you to be able to turn an immutable ref back into a mutable one. Not sure but I'm wondering if merge-op could help by letting you just merge new diffs onto refs instead of re-opening them as mutable in order to modify them (probably with periodic flattening of "merge chains" when using non-native snapshotters). But that's all too much for right now, so let me know what you think of the new approach here.